File: apps/frontend/rgbank/src/models/Thread.ts
import type { ExpenseStatus } from '@/models/General'
import type Student from '@/models/Student'

export type MessageType = 'STUDENT' | 'SYSTEM'

export interface Thread {
  thread_id: string
  messages?: Message[]
  unread_messages?: Message[]
}

export interface Message {
  message_id: string
  content: string
  created_at: string
  read_at?: string
  message_type: MessageType
  sender?: Student
  previous_status?: ExpenseStatus
  new_status?: ExpenseStatus
}

File: apps/backend/core-api/dto/apps/rgbank/thread.py
import msgspec

from dto.core.student import StudentDTO


class MessageDTO(msgspec.Struct):
    message_id: str
    content: str
    created_at: str | None = None
    read_at: str | None = None
    previous_status: str | None = None
    new_status: str | None = None
    sender: StudentDTO | None = None
    message_type: str | None = None


class AddMessageDTO(msgspec.Struct):
    message: str


class ThreadDTO(msgspec.Struct):
    thread_id: str
    messages: list[MessageDTO] | None = []
    unread_messages: list[MessageDTO] | None = []
